src/main/scala/uk/gov/nationalarchives/tdr/api/service/FileService.scala

```diff
@@ -54,15 +53,9 @@ class FileService(
 
     val row: (UUID, String, String) => FilemetadataRow = FilemetadataRow(uuidSource.uuid, _, _, now, userId, _)
     val rows: Future[List[Rows]] = customMetadataPropertiesRepository.getCustomMetadataValuesWithDefault.map(filePropertyValue => {
-      val allNodes = allEmptyDirectoryNodes ++ allFileNodes
-      val allNodesWithReference = if (referenceGeneratorFeatureBlock) {
-        allNodes.keys.zip(allNodes.values.zip(List.fill(allNodes.size)(None))).toMap
-      } else {
-        val generatedReferences = referenceGeneratorService.getReferences(allNodes.size)
-        allNodes.keys.zip(allNodes.values.zip(generatedReferences.map(Some(_)))).toMap
-      }
-      (allNodesWithReference map { case (path, (treeNode, reference)) =>
+      ((allEmptyDirectoryNodes ++ allFileNodes) map { case (path, treeNode) =>
         val parentId = treeNode.parentPath.map(path => allFileNodes.getOrElse(path, allEmptyDirectoryNodes(path)).id)
+        val parentFileReference = treeNode.parentPath.flatMap(path => allFileNodes.getOrElse(path, allEmptyDirectoryNodes(path)).reference)
```

Comment:
I would get the whole parent node first, then retrieve its id and reference separately. Something like this:

`val parentNode: Option[TreeNode] = treeNode.parentPath.map(path => allFileNodes.getOrElse(path, allEmptyDirectoryNodes(path)))`

Saves searching all the nodes twice to get the parent node.

Reply:
Good idea, I've made the changes.